### ISSUE TITLE
PushMeterRegistry implementations can silently stop sending metrics 

### ISSUE THREAD

**Issue by shakuzen** [micrometer-metrics/micrometer#1228]
See https://github.com/micrometer-metrics/micrometer/issues/1001#issuecomment-463023832 for what prompted me to create this issue. I see that previously we guarded against this in push-based registries at the time in cfbf1b0cbddc4575332218c25a7268b82cde3bf9. Now that we have the `PushMeterRegistry` abstraction, perhaps it is better that we apply this guard at that level, so it is consistent and new registry implementations do not miss this feature.

The risk of not having this is that when an uncaught exception happens, it is not logged and the `publish` task will no longer be scheduled. This makes it hard to notice/investigate these bugs. However, in cases like the issue in #1001, every publish would throw this error and it would never stop (nor would any metrics be published) since there was a bug in the logic preparing the data to publish. Logging a stack trace every publish forever seems a bit annoying for a user affected by this waiting for a fixed version (though they could update their setup to not enable/configure the registry once they realize this issue). Maybe there is a balance we could strike of catching and logging for this while not being too noisy?

**Comment by juliannzta** [comment]
I 100% agree on this and ideally should be high priority. The risk of silently losing monitoring far outweighs the annoyance of logging an error every time the scheduler runs. Quitting the schedule means no recovery other than an app restart and as a result application monitoring is unavailable.

Perhaps to avoid too much log noise, it would be useful to log a ERROR or WARN if without the full stacktace, but also log the entire stacktrace with some informative information on a DEBUG level. Additional configuration options would also be helpful for this as you mention.

#1001  is an example of how that situation can make this library unusable for a production environment. I (and others as noted on that ticket) spent considerable time scratching heads and debugging through code to get any understanding of why metrics aren't being published. As a sort of side not, the statsd registry has a similar issue, it logs nothing on success or failure. I still never got to the bottom of why it never worked in my production environment, I eventually changed approach.

Thanks of your great work on this library!!

**Comment by PyvesB** [comment]
We're using the CloudWatch implementation and suddenly observed that the metrics stopped being published. As the CloudWatch implementation relies on `PushMeterRegistry`, may the problem also be related to the issue described here?

Thanks for looking into this!

**Comment by shakuzen** [comment]
I have spent substantial time looking at the exception handling and differences between registry implementations. I think changing that is bigger scope than what should be tackled here, in a patch release. I have opened #1342 as the fix for the underlying issue here: `publish` can throw uncaught exceptions which aren't logged. Uncaught exceptions happening is almost certainly a bug in Micrometer code which we need to fix, but this hopefully gives users the info they need to report so we can fix it.
